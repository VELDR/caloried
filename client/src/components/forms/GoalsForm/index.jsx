import PropTypes from 'prop-types';
import { useState } from 'react';
import { injectIntl } from 'react-intl';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import toast from 'react-hot-toast';
import gain from '@static/images/gain.svg';
import lose from '@static/images/lose.svg';
import maintain from '@static/images/maintain.svg';
import FormFooter from '@components/forms/FormFooter';
import FormHeader from '@components/forms/FormHeader';

import { nextStep, setGoal } from '@pages/SignUp/actions';
import { selectGoal } from '@pages/SignUp/selectors';

import classes from './style.module.scss';

const GoalsForm = ({ goal, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const [selectedGoal, setSelectedGoal] = useState(goal);
  const goals = {
    lose: { label: formatMessage({ id: 'app_lose' }), image: lose },
    maintain: { label: formatMessage({ id: 'app_maintain' }), image: maintain },
    gain: { label: formatMessage({ id: 'app_gain' }), image: gain },
  };

  const handleGoalSelect = (newGoal) => {
    setSelectedGoal(newGoal);
  };

  const handleFormSubmit = () => {
    if (!selectedGoal) {
      toast.error(formatMessage({ id: 'app_error_goal' }));
    } else {
      dispatch(setGoal(selectedGoal));
      dispatch(nextStep());
    }
  };

  return (
    <div className={classes.form}>
      <FormHeader
        title={formatMessage({ id: 'app_goal_title' })}
        description={formatMessage({ id: 'app_goal_description' })}
      />

      <div className={classes.goal}>
        {Object.entries(goals).map(([key, { label, image }]) => (
          <div
            key={key}
            className={`${classes.goal__item} ${selectedGoal === key ? classes.active : ''}`}
            onClick={() => handleGoalSelect(key)}
          >
            <img className={classes.image} src={image} alt={label} />
            <div className={classes.label}>{label}</div>
          </div>
        ))}
      </div>

      <div className={classes.footer}>
        <FormFooter onContinue={handleFormSubmit} />
      </div>
    </div>
  );
};

GoalsForm.propTypes = {
  goal: PropTypes.string,
  intl: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  goal: selectGoal,
});

export default injectIntl(connect(mapStateToProps)(GoalsForm));
