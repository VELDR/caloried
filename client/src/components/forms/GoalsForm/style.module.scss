@import '@styles/base';

.form {
  @include flex-center-column(2rem);
  justify-content: space-between;
  text-align: center;
  height: 100%;

  .goal {
    @include flex-center(1rem);
    flex-wrap: wrap;

    &__item {
      cursor: pointer;
      @include flex-center-column(1rem);
      padding: 1rem;
      border-radius: 1rem;
      border: 3px solid var(--color-text-secondary);
      @include box(8rem, 10rem);
      overflow: hidden;

      &:hover {
        background-color: var(--color-primary);
        border-color: var(--color-primary);
      }
      &:hover .label {
        font-weight: 700;
        color: var(--color-text-primary);
      }
      &:hover .image {
        transform: scale(1.25);
      }
      .image {
        @include box(8rem);
        transition: all 0.5s ease-in-out;
      }
      .label {
        text-align: center;
        color: var(--color-text-secondary);
        font-size: $font-size-regular;
      }
    }
    .active {
      background-color: var(--color-primary);
      border-color: var(--color-text-primary);
      .label {
        font-weight: 700;
        color: var(--color-text-primary);
      }
      .image {
        transform: scale(1.2);
      }
    }
  }

  .footer {
    width: 100%;
  }
}
