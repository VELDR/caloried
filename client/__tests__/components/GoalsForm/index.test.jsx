import GoalsForm from '@components/forms/GoalsForm';
import { fireEvent, render } from '@utils/testHelper';

let wrapper;

const mockProps = {
  goal: 'maintain',
  intl: { formatMessage: jest.fn() },
};

beforeEach(() => {
  wrapper = render(<GoalsForm {...mockProps} />);
});

describe('GoalsForm Component', () => {
  it('should render the component correctly', () => {
    const { getByTestId, getByText } = wrapper;

    expect(getByTestId('goals-form')).toBeInTheDocument();
    expect(getByTestId('form-header')).toBeInTheDocument();

    expect(getByText('Turunkan Berat Badan')).toBeInTheDocument();
    expect(getByText('Pertahankan Berat Badan')).toBeInTheDocument();
    expect(getByText('Menambah Berat Badan')).toBeInTheDocument();
  });
  it('should handle goal selection', () => {
    const { getByText, getByTestId } = wrapper;

    fireEvent.click(getByText('Menambah Berat Badan'));
    expect(getByTestId('goal-item-gain')).toHaveClass('active');
    expect(getByTestId('goal-item-maintain')).not.toHaveClass('active');
  });
  it('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
