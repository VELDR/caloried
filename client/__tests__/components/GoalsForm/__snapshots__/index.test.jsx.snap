// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoalsForm Component should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form"
        data-testid="goals-form"
      >
        <div
          class="header"
          data-testid="form-header"
        >
          <div
            class="header__title"
          >
            Tetapkan Tujuan Kebugaran Anda
          </div>
          <div
            class="header__description"
          >
            Beri tahu kami apa yang menjadi tujuan Anda
          </div>
        </div>
        <div
          class="goal"
        >
          <div
            class="goal__item "
            data-testid="goal-item-lose"
          >
            <img
              alt="Turunkan Berat Badan"
              class="image"
            />
            <div
              class="label"
            >
              Turunkan Berat Badan
            </div>
          </div>
          <div
            class="goal__item "
            data-testid="goal-item-maintain"
          >
            <img
              alt="Pertahankan Berat Badan"
              class="image"
            />
            <div
              class="label"
            >
              Pertahankan Berat Badan
            </div>
          </div>
          <div
            class="goal__item "
            data-testid="goal-item-gain"
          >
            <img
              alt="Menambah Berat Badan"
              class="image"
            />
            <div
              class="label"
            >
              Menambah Berat Badan
            </div>
          </div>
        </div>
        <div
          class="footer"
        >
          <div
            class="footer"
            data-testid="form-footer"
          >
            <button
              class="button footer__right footer__button"
              data-testid="primary-button"
              type="submit"
            >
              Mulai
               
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EastIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m15 5-1.41 1.41L18.17 11H2v2h16.17l-4.59 4.59L15 19l7-7-7-7z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form"
      data-testid="goals-form"
    >
      <div
        class="header"
        data-testid="form-header"
      >
        <div
          class="header__title"
        >
          Tetapkan Tujuan Kebugaran Anda
        </div>
        <div
          class="header__description"
        >
          Beri tahu kami apa yang menjadi tujuan Anda
        </div>
      </div>
      <div
        class="goal"
      >
        <div
          class="goal__item "
          data-testid="goal-item-lose"
        >
          <img
            alt="Turunkan Berat Badan"
            class="image"
          />
          <div
            class="label"
          >
            Turunkan Berat Badan
          </div>
        </div>
        <div
          class="goal__item "
          data-testid="goal-item-maintain"
        >
          <img
            alt="Pertahankan Berat Badan"
            class="image"
          />
          <div
            class="label"
          >
            Pertahankan Berat Badan
          </div>
        </div>
        <div
          class="goal__item "
          data-testid="goal-item-gain"
        >
          <img
            alt="Menambah Berat Badan"
            class="image"
          />
          <div
            class="label"
          >
            Menambah Berat Badan
          </div>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="footer"
          data-testid="form-footer"
        >
          <button
            class="button footer__right footer__button"
            data-testid="primary-button"
            type="submit"
          >
            Mulai
             
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="EastIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m15 5-1.41 1.41L18.17 11H2v2h16.17l-4.59 4.59L15 19l7-7-7-7z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
